File: src/providers/openai-cua/rules/no-blind-safety-check-ack.ts
/**
 * Flags a `.filter()` over a safety-checks array whose callback never
 * references `.code` or `.message` — i.e. it acknowledges every check
 * present instead of evaluating each one against a policy.
 */
const rule = {
  meta: {
    type: 'suggestion',
    docs: {
      description: 'Acknowledging pending safety checks must evaluate each check, not blanket-pass them',
      category: 'correctness',
      rationale:
        '`acknowledged_safety_checks` exists specifically so a developer can selectively confirm checks they have actually reviewed — each entry carries an id/code/message. A filter that only checks the row\'s shape (e.g. that it is an object) and never inspects `.code` or `.message` acknowledges every check present, defeating the purpose of the field entirely.',
      docsUrl: 'https://developers.openai.com/api/docs/guides/tools-computer-use',
      recommended: true,
    },
    messages: {
      blindAck:
        'This filter never inspects .code or .message on each safety check — it acknowledges every check present instead of evaluating it against a policy.',
    },
  },
  create(context: any) {
    function propName(node: any): string | undefined {
      if (!node) return undefined;
      if (node.type === 'Identifier') return node.name;
      if (node.type === 'Literal' && typeof node.value === 'string') return node.value;
      return undefined;
    }

    function sourceLooksLikeSafetyChecks(node: any): boolean {
      if (!node) return false;
      const name = propName(node) ?? (node.type === 'MemberExpression' ? propName(node.property) : undefined);
      if (name && /safety.?check/i.test(name)) return true;
      if (node.type === 'LogicalExpression') {
        return sourceLooksLikeSafetyChecks(node.left) || sourceLooksLikeSafetyChecks(node.right);
      }
      if (node.type === 'MemberExpression') {
        return sourceLooksLikeSafetyChecks(node.object) || sourceLooksLikeSafetyChecks(node.property);
      }
      return false;
    }

    function referencesCodeOrMessage(node: any, depth = 0): boolean {
      if (!node || typeof node !== 'object' || depth > 10) return false;
      if (Array.isArray(node)) return node.some((n) => referencesCodeOrMessage(n, depth + 1));
      if (node.type === 'MemberExpression') {
        const name = propName(node.property);
        if (name === 'code' || name === 'message') return true;
      }
      for (const key of Object.keys(node)) {
        if (key === 'parent' || key === 'loc' || key === 'range') continue;
        const val = (node as any)[key];
        if (val && typeof val === 'object') {
          if (referencesCodeOrMessage(val, depth + 1)) return true;
        }
      }
      return false;
    }

    return {
      CallExpression(node: any) {
        const callee = node.callee;
        if (callee?.type !== 'MemberExpression' || callee.property?.name !== 'filter') return;
        if (!sourceLooksLikeSafetyChecks(callee.object)) return;

        const fn = node.arguments?.[0];
        if (fn?.type !== 'ArrowFunctionExpression' && fn?.type !== 'FunctionExpression') return;

        if (!referencesCodeOrMessage(fn.body)) {
          context.report({ node, messageId: 'blindAck' });
        }
      },
    };
  },
};

export const openaiCuaNoBlindSafetyCheckAckRule = rule;

File: src/providers/openai-cua/rules/python/no_blind_safety_check_ack.py
"""openai-cua-no-blind-safety-check-ack

Parity with the JS rule: flags a comprehension/filter over a safety-checks
collection whose predicate never inspects `.code`/`.message` (or the dict-key
equivalents) — i.e. it acknowledges every check present instead of evaluating
each one against a policy.
"""
from __future__ import annotations

import ast
import re
import sys
from pathlib import Path

from ast_utils import loc

_PROVIDER = Path(__file__).resolve().parents[2]
if str(_PROVIDER) not in sys.path:
    sys.path.insert(0, str(_PROVIDER))

RULE_KEY = "openai-cua-no-blind-safety-check-ack"

SAFETY_CHECK_NAME_RE = re.compile(r"safety[_-]?check", re.I)
CODE_OR_MESSAGE_KEYS = frozenset({"code", "message"})


def _looks_like_safety_checks(node: ast.AST | None) -> bool:
    if node is None:
        return False
    if isinstance(node, ast.Name):
        return bool(SAFETY_CHECK_NAME_RE.search(node.id))
    if isinstance(node, ast.Attribute):
        return bool(SAFETY_CHECK_NAME_RE.search(node.attr)) or _looks_like_safety_checks(node.value)
    if isinstance(node, ast.BoolOp):
        return any(_looks_like_safety_checks(v) for v in node.values)
    return False


def _references_code_or_message(node: ast.AST) -> bool:
    for n in ast.walk(node):
        if isinstance(n, ast.Attribute) and n.attr in CODE_OR_MESSAGE_KEYS:
            return True
        if isinstance(n, ast.Subscript):
            sl = n.slice
            if isinstance(sl, ast.Constant) and sl.value in CODE_OR_MESSAGE_KEYS:
                return True
        if isinstance(n, ast.Call) and isinstance(n.func, ast.Attribute) and n.func.attr == "get":
            if n.args and isinstance(n.args[0], ast.Constant) and n.args[0].value in CODE_OR_MESSAGE_KEYS:
                return True
    return False


def check(tree: ast.AST, path: str, source: str) -> list[dict]:
    out: list[dict] = []

    for node in ast.walk(tree):
        if isinstance(node, ast.ListComp) and len(node.generators) == 1:
            gen = node.generators[0]
            if not gen.ifs or not _looks_like_safety_checks(gen.iter):
                continue
            if any(_references_code_or_message(cond) for cond in gen.ifs):
                continue
            line, col, _, _ = loc(node)
            out.append({"file": path, "line": line, "column": col, "ruleKey": RULE_KEY})

        elif isinstance(node, ast.Call) and isinstance(node.func, ast.Name) and node.func.id == "filter":
            if len(node.args) != 2:
                continue
            fn_arg, iterable = node.args
            if not isinstance(fn_arg, ast.Lambda):
                continue
            if not _looks_like_safety_checks(iterable):
                continue
            if _references_code_or_message(fn_arg.body):
                continue
            line, col, _, _ = loc(node)
            out.append({"file": path, "line": line, "column": col, "ruleKey": RULE_KEY})

    return out
